### DISCUSSION TITLE
Keycloak Organizations Feedback - Keep Group API working for organizations

### DISCUSSION THREAD

**Discussion by jbman** [keycloak/keycloak#31470] [Feedback]
We have implemented and still evolve extensions around Keycloak groups, so that group admins can invite users and manage roles of users in their group. Keycloak (version 25.0.1) organizations seem to go in a similar direction. However, I found some constraints which make it hard to get our extensions and the new Keycloak Organizations working together.  

**Allow management of organizations with existing groups API**
The new organization feature in Keycloak creates a group if you create an organization. However the Admin API endpoints refuse to work with the group which results in errors in the Admin Console when you try to open the group ("Network response was not OK.").
The API responds with 400 and error message "Cannot manage organization related group via non Organization API."

The groups API should still work for organization, so that existing UIs and applications can work with a organization as a group.
Organizations may bring additional constraints and additional features but should still be accessible through established APIs.

**Allow sending invitation without the need for user self registration**
When invitation is used, there is no invitation email send if self-registration is disabled. But it should be possible to invite a user so that the users roles can be managed right away. Finally the user logs in with their existing IdP account and have all roles they need for their work.

**Allow a user to be member of multliple groups**
We also started with the constraint, that a user can be member of one group, only. But this is inconvenient, as users need to start using separate accounts. So we have started implementing a group selection after login. This way a user who is member of multiple groups can decide for one group after login and work in the context of this group in a session.

Did you already discuss or decide any of the above points? Happy to read your position and plans on those.

**Comment by jbman** [comment]
I found that https://github.com/keycloak/keycloak/issues/31438 is related , where you plan to add an organization selection.

**Comment by pedroigor** [comment]
> We have implemented and still evolve extensions around Keycloak groups, so that group admins can invite users and manage roles of users in their group. Keycloak (version 25.0.1) organizations seem to go in a similar direction. However, I found some constraints which make it hard to get our extensions and the new Keycloak Organizations working together.
> 
> **Allow management of organizations with existing groups API** The new organization feature in Keycloak creates a group if you create an organization. However the Admin API endpoints refuse to work with the group which results in errors in the Admin Console when you try to open the group ("Network response was not OK."). The API responds with 400 and error message "Cannot manage organization related group via non Organization API."
> 
> The groups API should still work for organization, so that existing UIs and applications can work with a organization as a group. Organizations may bring additional constraints and additional features but should still be accessible through established APIs.

In upstream, you won't even see/get the "organization groups" from the Admin UI/APIs.

The reason why we are hiding these groups is because they are really an implementation detail that people should not be aware of. Any capability that you need in an organization, even if granted for free by the fact we are using groups, should be done through the organization feature and not by using those groups directly.

For instance, we have in our [roadmap](https://github.com/keycloak/keycloak/issues/30180) the possibility of managing roles and groups in the context of an organization. 

> 
> **Allow sending invitation without the need for user self registration** When invitation is used, there is no invitation email send if self-registration is disabled. But it should be possible to invite a user so that the users roles can be managed right away. Finally the user logs in with their existing IdP account and have all roles they need for their work.

You are only forced to enable self-registration when inviting new users. Inviting existing users will not require enabling this setting.

Can you please elaborate on what you mean by "user roles can be managed right away"?

> 
> **Allow a user to be member of multliple groups** We also started with the constraint, that a user can be member of one group, only. But this is inconvenient, as users need to start using separate accounts. So we have started implementing a group selection after login. This way a user who is member of multiple groups can decide for one group after login and work in the context of this group in a session.

We are enabling joining multiple accounts here https://github.com/keycloak/keycloak/pull/29613. Should be merged soon.

What we don't have clear/planned yet is the possibility to select organizations when authenticating. Is it related to https://github.com/keycloak/keycloak/issues/31438?

> 
> Did you already discuss or decide any of the above points? Happy to read your position and plans on those.

**Comment by jbman** [comment]
Thanks for your explanations @pedroigor. 

By "user roles can be managed right away" I meant that a user can be invited so that the account is created and roles can be added to this account.
So it seems like currently a multi-step process is required to reach this: create the user, add user to organization, write an email manually to the user that he could log in. 
But maybe it is already improved or planned, that a user can be invited to make this less steps for the inviting admin.

Selecting an organization is related to https://github.com/keycloak/keycloak/issues/31438. 
I saw that you're discussing how the client could select the organization. For our client apps, they would rather delegate selection to Keycloak (not specific the org on their own). They should be able to request that they need one specific organization to work correctly. Then they can let the user do tasks for just this organization. They also need to provide link for the user to switch organization, which would be a re-authentication at Keycloak to select a new organization context in the current session.
